test/storage/garbage_collector_test.cpp

```diff
@@ -671,4 +671,894 @@ TEST_F(GarbageCollectorTests, InsertUpdate1) {
+    transaction::TransactionManager txn_manager(&buffer_pool_, true, LOGGING_DISABLED);
+    GarbageCollectorDataTableTestObject tested(&block_store_, max_columns_, &generator_);
+    storage::GarbageCollector gc(&txn_manager);
+
+    // T1
+    auto *txn1 = txn_manager.BeginTransaction();
+
+    // INSERT
+    auto *txn_insert = txn_manager.BeginTransaction();
+    auto *insert_tuple = tested.GenerateRandomTuple(&generator_);
+    storage::TupleSlot slot = tested.table_.Insert(txn_insert, *insert_tuple);
+    txn_manager.Commit(txn_insert, TestCallbacks::EmptyCallback, nullptr);
+
+    // DELETE
+    auto *txn_delete = txn_manager.BeginTransaction();
+    tested.table_.Delete(txn_delete, slot);
+    txn_manager.Commit(txn_delete, TestCallbacks::EmptyCallback, nullptr);
+
+    // Active txn should not see any version
+    tested.SelectIntoBuffer(txn1, slot);
+    EXPECT_FALSE(tested.select_result_);
+
+    // Nothing should be collected
+    EXPECT_EQ(std::make_pair(0u, 0u), gc.PerformGarbageCollection());
+
+    txn_manager.Commit(txn1, TestCallbacks::EmptyCallback, nullptr);
+    // Both DELETE and INSERT should be unlinked
+    // T1 (read only) should be unlinked and deallocated
+    EXPECT_EQ(std::make_pair(0u, 3u), gc.PerformGarbageCollection());
+
+    // DELETE and INSERT should be deallocated
+    EXPECT_EQ(std::make_pair(2u, 0u), gc.PerformGarbageCollection());
+
+    // Nothing should be unlinked or deallocated
+    EXPECT_EQ(std::make_pair(0u, 0u), gc.PerformGarbageCollection());
+  }
+}
+
+/*
+ * Consider performing GC on the version chain sorted from newest to oldest:
+ * DELETE
+ * T1 <- active txn
+ * INSERT
+ */
+// NOLINTNEXTLINE
+TEST_F(GarbageCollectorTests, DeleteTest2) {
+  for (uint32_t iteration = 0; iteration < num_iterations_; ++iteration) {
+    transaction::TransactionManager txn_manager(&buffer_pool_, true, LOGGING_DISABLED);
+    GarbageCollectorDataTableTestObject tested(&block_store_, max_columns_, &generator_);
+    storage::GarbageCollector gc(&txn_manager);
+
+    // INSERT
+    auto *txn_insert = txn_manager.BeginTransaction();
+    auto *insert_tuple = tested.GenerateRandomTuple(&generator_);
+    storage::TupleSlot slot = tested.table_.Insert(txn_insert, *insert_tuple);
+    txn_manager.Commit(txn_insert, TestCallbacks::EmptyCallback, nullptr);
+
+    // T1
+    auto *txn1 = txn_manager.BeginTransaction();
+
+    // DELETE
+    auto *txn_delete = txn_manager.BeginTransaction();
+    tested.table_.Delete(txn_delete, slot);
+    txn_manager.Commit(txn_delete, TestCallbacks::EmptyCallback, nullptr);
+
+    // Active txn should see correct version
+    auto *select_tuple = tested.SelectIntoBuffer(txn1, slot);
+    EXPECT_TRUE(tested.select_result_);
+    EXPECT_TRUE(StorageTestUtil::ProjectionListEqualShallow(tested.Layout(), select_tuple, insert_tuple));
+
+    // INSERT should be unlinked
+    EXPECT_EQ(std::make_pair(0u, 1u), gc.PerformGarbageCollection());
+
```

Comment:
I suggest you add an ``EXPECT_EQ(std::make_pair(0u, 0u), gc.PerformGarbageCollection());`` here to test if INSERT is deallocated.

Reply:
I think that [the next `EXPECT_EQ`](https://github.com/cmu-db/terrier/pull/340/files/e301b972d160510ff76929af06ef02088b12a170#diff-8da770a88e7232c376d2a3a1ec56f9dbR1265) does test that